alws/utils/pulp_utils.py

```diff
@@ -9,13 +9,83 @@
     CoreContent,
     CoreRepository,
     CoreRepositoryContent,
+    RpmModulemd,
+    RpmModulemdPackages,
     RpmPackage,
 )
 
 
 def get_uuid_from_pulp_href(pulp_href: str) -> uuid.UUID:
     return uuid.UUID(pulp_href.split("/")[-2])
 
+# TODO: Merge with get_rpm_packages_from_repository
+# or keep separate functions but make them share code
+def get_rpm_module_packages_from_repository(
+    repo_id: uuid.UUID,
+    module: str,
+    pkg_names: typing.List[str] = None,
+    pkg_versions: typing.List[str] = None,
+    pkg_epochs: typing.List[str] = None,
+) -> typing.List[RpmPackage]:
+    module_name, module_stream = module.split(":")
+    first_subq = (
+        select(CoreRepository.pulp_id)
+        .where(CoreRepository.pulp_id == repo_id)
+        .scalar_subquery()
+    )
+    second_subq = (
+        select(CoreRepositoryContent.content_id)
+        .where(
+            CoreRepositoryContent.repository_id == first_subq,
+            CoreRepositoryContent.version_removed_id.is_(None),
+        )
+        .scalar_subquery()
+    )
+    third_subq = (
+        select(CoreContent.pulp_id)
+        .where(
+            CoreContent.pulp_id.in_(second_subq),
+            CoreContent.pulp_type == "rpm.modulemd",
+        )
+        .scalar_subquery()
+    )
+    # By default, it should return the modules ordered
+    # by version, or that's what I observed so far.
+    # If it's not the case, add order_by(RpmModulemd.version)
+    # to this subquery.
+    # We want to ensure that the first mapping package is
+    # taken as the right one later on at get_matching_albs_packages
+    fourth_subq = (
+        select(RpmModulemd.content_ptr_id)
+        .where(
+            RpmModulemd.content_ptr_id.in_(third_subq),
+            RpmModulemd.name == module_name,
+            RpmModulemd.stream == module_stream,
+        )
+        .scalar_subquery()
+    )
+    last_subq = (
+        select(RpmModulemdPackages.package_id)
+        .where(
+            RpmModulemdPackages.modulemd_id.in_(fourth_subq),
+        )
+        .scalar_subquery()
+    )
```

Comment:
are you sure that `RpmModulemdPackages's` exists even if we don't provide artifacts/packages during module creation?

Reply:
Right, I've double checked this, and what I observed is that, when processing a build after is finished, we do add artifacts/packages, which results in having 2 modules in pulp:
1. One that is created upon the build creation
2. The other one that is created when the build is finished

i.e. look at [this build](https://build.almalinux.org/build/5781), here's the [link to the modules.yaml](https://build.almalinux.org/pulp/content/builds/AlmaLinux-8-x86_64-5781-br/repodata/87c24e46d7e2adbcca25f2919b924790bde613f5ce60d857faffe1ed5123b00e-modules.yaml) file of that resulting build repo.

Reply:
Finally, we're not taking the modules data from pulp's db but from production modules.yaml files